### ISSUE TITLE
Sometimes Vector gets no presence data at all

### ISSUE THREAD

**Issue by ara4n** [vector-im/vector-web#186]
...and the whole membership list appears in state 'unknown'.  Only seen this once.

**Comment by ara4n** [comment]
Just repro'd it again. Nothing obvious in the JS console. Will try to keep the tab open for good luck.

**Comment by dbkr** [comment]
Just reproed this: got only my presence in the initialsync. Filed in synapse as SYN-523.

**Comment by ara4n** [comment]
apparently fixed!

**Comment by dbkr** [comment]
Apparently not - just got this bug again with the newly redeployed synapse on matrix.org (and syn-523 is still open).

**Comment by ara4n** [comment]
Still happening even with all-new-erik-presence:

![screen shot 2016-02-25 at 00 56 35](https://cloud.githubusercontent.com/assets/1294269/13305330/ca1c08d0-db5a-11e5-88f9-282312c50f9a.png)

**Comment by erikjohnston** [comment]
Is this after its been open a while? Or after just opening the app?

**Comment by ara4n** [comment]
I've seen it when opening the app in the past (i.e. on initial /sync), 
but more recently i've seen it a few times when alt-tabbing to an 
existing session.

Hmmmmmmmm. is this just going to be people restarting the server 
nowadays and losing all presence state?

**Comment by erikjohnston** [comment]
Exciting. Presence states should persist across server restarts nowadays.

**Comment by ara4n** [comment]
Ooooh!!! Very cool :D  Then I have no idea what's causing presence to go 
missing.  I saw it last a few days ago.  What would be the preferred 
means of debugging next time it happens?

**Comment by erikjohnston** [comment]
Not sure really, the stream token might be useful?

**Comment by ara4n** [comment]
Reproduced again just now.  I'm sure presence was working when I slept my box at 18:15, but on unsleeping at 23:47 everyone was shown as 'unknown'.  Meanwhile I actually had the network inspector running for once.  The console says:

```
2016-03-03 18:10:38.458 Presence.js:83 Presence: online
2016-03-03 18:14:42.561 sync.js:499 /sync error Error: CORS request rejected: https://matrix.org/_matrix/client/r0/sync?filter=21&timeout=30000&since=s14…gzNzA2MDM1CjAwMmZzaWduYXR1cmUgFvDijWZSGyig_37n17-5KQqVz4jmYV1kkazLihMp6FwK
    at i (https://vector.im/develop/bundle.js:147:22289)
    at XMLHttpRequest.r (https://vector.im/develop/bundle.js:147:22099)(anonymous function) @ sync.js:499i @ q.js:844(anonymous function) @ q.js:870T.g.when @ q.js:1122r.promiseDispatch @ q.js:788(anonymous function) @ q.js:604r @ q.js:137e @ q.js:125(anonymous function) @ main.js:64i.run @ browser.js:64r @ browser.js:34
2016-03-03 18:14:42.563 sync.js:500 Error: CORS request rejected: https://matrix.org/_matrix/client/r0/sync?filter=21&timeout=30000&since=s14…gzNzA2MDM1CjAwMmZzaWduYXR1cmUgFvDijWZSGyig_37n17-5KQqVz4jmYV1kkazLihMp6FwK(…)(anonymous function) @ sync.js:500i @ q.js:844(anonymous function) @ q.js:870T.g.when @ q.js:1122r.promiseDispatch @ q.js:788(anonymous function) @ q.js:604r @ q.js:137e @ q.js:125(anonymous function) @ main.js:64i.run @ browser.js:64r @ browser.js:34
2016-03-03 18:14:42.563 sync.js:49 Starting keep-alive
2016-03-03 18:14:42.566 MatrixChat.js:554 MatrixClient sync state => ERROR
2016-03-03 18:14:42.674 MatrixChat.js:554 MatrixClient sync state => SYNCING
2016-03-03 23:57:51.595 sync.js:499 /sync error e.exports.MatrixError(anonymous function) @ sync.js:499i @ q.js:844(anonymous function) @ q.js:870T.g.when @ q.js:1122r.promiseDispatch @ q.js:788(anonymous function) @ q.js:604r @ q.js:137e @ q.js:125(anonymous function) @ main.js:64i.run @ browser.js:64r @ browser.js:34
2016-03-03 23:57:51.596 sync.js:500 e.e…s.MatrixError {errcode: "ORG.MATRIX.JSSDK_TIMEOUT", name: "ORG.MATRIX.JSSDK_TIMEOUT", message: "Locally timed out waiting for a response", data: Object}(anonymous function) @ sync.js:500i @ q.js:844(anonymous function) @ q.js:870T.g.when @ q.js:1122r.promiseDispatch @ q.js:788(anonymous function) @ q.js:604r @ q.js:137e @ q.js:125(anonymous function) @ main.js:64i.run @ browser.js:64r @ browser.js:34
2016-03-03 23:57:51.597 sync.js:49 Starting keep-alive
2016-03-03 23:57:51.600 MatrixChat.js:554 MatrixClient sync state => ERROR
2016-03-03 23:57:58.957 MatrixChat.js:554 MatrixClient sync state => SYNCING```

But most interestingly, filtering the network requests on /sync, I see two failed requests which match the timestamps of the errors in the JS logs.  After the first one, we see a /sync with timeout=0 which returns no results at all (which is unsurprising, given the first outage is a brief one)

```
{
    "next_batch": "s14361899_390700_6174_804055_4455",
    "account_data": {
        "events": []
    },
    "rooms": {
        "leave": {},
        "join": {},
        "invite": {}
    },
    "presence": {
        "events": []
    }
}
```

...and then after the second one, it returns 411 presence updates:

```
    "presence": {
        "events": [{
            "content": {
                "last_active_ago": 12726383,
                "presence": "offline"
            },
            "type": "m.presence",
            "sender": "@freenode_arnsholt:matrix.org"
        }, {
            "content": {
                "last_active_ago": 14332409,
                "presence": "offline"
            },
            "type": "m.presence",
            "sender": "@freenode_eschmidbauer:matrix.org"
        }, {
```

...as well as all the gappy sync room deltas.

So, from the synapse perspective, everything looks great.  My hunch is that the js-sdk is simply dropping all of the presence state after a failed /sync request or timeout.

@kegsay, does this ring a bell? (not asking you to fix it, just if you recall the impl).  re-assigning to @dbkr (or me, depending on who gets to it first)

**Comment by Kegsay** [comment]
Your idea makes sense but I don't think that can actually happen.

Looking at https://github.com/matrix-org/matrix-js-sdk/blob/develop/lib/sync.js#L434 ,`_processSyncResponse` is *always* called if the HTTP request was successful. That *always* then results in https://github.com/matrix-org/matrix-js-sdk/blob/develop/lib/sync.js#L530 which updates the `User` and emits an event.

It's plausible that the underlying HTTP layer may knife the connection early during retries, but this would then result in a rejection, which should then retry with the *same token* and so should *always* eventually get those presence events.

To put it another way, we aren't treating presence in any special way compared to actual messages, so if your theory was correct then we should be seeing missing messages/members/other events but we don't. This makes me think that the JS SDK *is actually updating* but this is not then triggering a re-render of the right bits in React-land?

**Comment by ara4n** [comment]
Hm... I have also seen (very rarely) scrollback getting lost: https://github.com/vector-im/vector-web/issues/981.

It's possible that JS & React have got out of sync - next time I see it i'll go poking in the depths of the JS SDK's state to check.  I saw it a few minutes ago but subsequently seems to have somehow fixed itself.

**Comment by Kegsay** [comment]
> I saw it a few minutes ago but subsequently seems to have somehow fixed itself.

This reinforces my theory that it is updating at the JS SDK layer but just not immediately triggering a re-render.

**Comment by ara4n** [comment]
I don't understand why the state would ever have gone to 'unknown' though...  and empirically presence events which have happened /since/ the hiatus tend to be displayed correctly.  Something is flushing all the presence state, either in JS or React, and dunno what.

**Comment by ara4n** [comment]
tentatively closing this in the hope that https://github.com/matrix-org/matrix-react-sdk/pull/196 fixes it
